kroxylicious-kms-providers/kroxylicious-kms-provider-azure-key-vault-kms/src/test/java/io/kroxylicious/kms/provider/azure/auth/OauthClientCredentialsTokenServiceTest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.provider.azure.auth;
+
+import java.net.URI;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.temporal.ChronoUnit;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import org.assertj.core.api.ThrowableAssertAlternative;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.kroxylicious.kms.httpmock.MockServer.Header;
+import io.kroxylicious.kms.httpmock.MockServer.Request;
+import io.kroxylicious.kms.httpmock.MockServer.Responder;
+import io.kroxylicious.kms.provider.azure.MalformedResponseBodyException;
+import io.kroxylicious.kms.provider.azure.UnexpectedHttpStatusCodeException;
+import io.kroxylicious.kms.provider.azure.config.auth.EntraIdentityConfig;
+import io.kroxylicious.proxy.config.secret.InlinePassword;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+import static io.kroxylicious.kms.httpmock.MockServer.getHttpServer;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class OauthClientCredentialsTokenServiceTest {
+
+    public static final String TOKEN = "abcdef";
+    public static final String TENANT_ID = "my-tenant";
+    public static final String CLIENT_ID = "id";
+    public static final String CLIENT_SECRET = "secret";
+
+    @CsvSource(value = { "https://scope/.default", "null" }, nullValues = "null")
+    @ParameterizedTest
+    void getToken(@Nullable String inputScope) {
+        Instant fixedInstant = Instant.now();
+        Clock clock = Clock.fixed(fixedInstant, ZoneId.of("UTC"));
+        int expirySeconds = 3200;
+        Responder resp = exchange -> {
+            var response = Map.of("token_type", "Bearer", "expires_in", expirySeconds, "access_token", TOKEN);
```

Comment:
When writing unit tests like these, I try to use the responses copied straight from the documentation of the services, with as few changes as possible, so we are testing against responses from an authoritative known good source.
So in this case, I'd copy the response from:

https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#successful-response-1

I think if you don't do this there's more risk that the code in your own misunderstandings of the API into the responses you hand-craft.

See io.kroxylicious.kms.provider.aws.kms.credentials.Ec2MetadataCredentialsProviderTest#KNOWN_GOOD_SECURITY_CREDENTIAL_RESPONSE

Reply:
sounds good, slight hassle that they've used ellipsis to reduce the size of the tokens and made their examples slightly less realistic.